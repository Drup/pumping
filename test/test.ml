type t = [%re? `A, Star (`B | `C)] ;;

(* Success *)
(`A `End : t) ;;
(`A (`B `End) : t) ;;
(`A (`B (`B `End)) : t) ;;
(`A (`C (`B `End)) : t) ;;

(* Failure *)
(`A : t) ;;
(`B `End : t) ;;
(`A (`A `End) : t) ;;
(`A (`C (`A `End)) : t) ;;


(* pattern syntaxes *)
(`A `End : [%re? 'A'..'Z']);;
(`C `End : [%re? Inter ((`A | `C),(`B | `C))]);;
(`C `End : [%re? (`A | `C) :: (`B | `C)]);;
(`End : [%re? Epsilon]);;

(* posix *)
(`a `End :[%re"a"]) ;;
(`abc `End :[%re"abc"]) ;;

(`a `End :[%re"[ab]"]) ;;
(`c `End :[%re"[ab]"]) ;; (*fail*)

(`a `End :[%re"[a-d]"]) ;;
(`x `End :[%re"[a-d]"]) ;; (*fail*)

(`a (`a (`a `End)) :[%re"a*"]) ;;
(`End :[%re"a*"]) ;;

(`a (`a (`a `End)) :[%re"a+"]) ;;
(`End :[%re"a+"]) ;; (*fail*)

(`a (`a (`a `End)) :[%re"a{3,}"]) ;;
(`a (`a (`a (`a `End))) :[%re"a{3,}"]) ;;
(`a (`a `End) :[%re"a{3,}"]) ;; (*fail*)

(`a (`a (`a `End)) :[%re"a{3,4}"]) ;;
(`a (`a (`a (`a `End))) :[%re"a{3,4}"]) ;;
(`a (`a `End) :[%re"a{3,4}"]) ;; (*fail*)
(`a (`a (`a (`a (`a `End)))) :[%re"a{3,4}"]) ;; (*fail*)

(`b `End :[%re"[ab]&[bc]"]) ;;
(`a `End :[%re"[ab]&[bc]"]) ;; (*fail*)


(* equivalences *)
let id x = x ;;
(id : [%re"aa*a"] -> [%re"a{2,}"]);;
(id : [%re"[ab]&[bc]"] -> [%re"b"]);;
(id : [%re"(a|b)(c|d)"] -> [%re"[a]c|[a]d|[b]c|[b]d"]);;

let f x = (x : [%re"a*"] :> [%re"[ab]*"]) ;;
