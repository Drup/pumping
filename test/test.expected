
type t = [ `A of [ `B of 'a | `C of 'a | `End ] as 'a ]
- : t = `A `End
- : t = `A (`B `End)
- : t = `A (`B (`B `End))
- : t = `A (`C (`B `End))
Characters 16-18:
  (`A : t) ;;
   ^^
Error: This expression has type [> `A ]
       but an expression was expected of type t
       Types for tag `A are incompatible
Characters 1-8:
  (`B `End : t) ;;
   ^^^^^^^
Error: This expression has type [> `B of [> `End ] ]
       but an expression was expected of type t
       The second variant type does not allow tag(s) `B
Characters 4-13:
  (`A (`A `End) : t) ;;
      ^^^^^^^^^
Error: This expression has type [> `A of [> `End ] ]
       but an expression was expected of type
         [ `B of 'a | `C of 'a | `End ] as 'a
       The second variant type does not allow tag(s) `A
Characters 8-17:
  (`A (`C (`A `End)) : t) ;;
          ^^^^^^^^^
Error: This expression has type [> `A of [> `End ] ]
       but an expression was expected of type
         [ `B of 'a | `C of 'a | `End ] as 'a
       The second variant type does not allow tag(s) `A
- : [ `A of [ `End ]
    | `B of [ `End ]
    | `C of [ `End ]
    | `D of [ `End ]
    | `E of [ `End ]
    | `F of [ `End ]
    | `G of [ `End ]
    | `H of [ `End ]
    | `I of [ `End ]
    | `J of [ `End ]
    | `K of [ `End ]
    | `L of [ `End ]
    | `M of [ `End ]
    | `N of [ `End ]
    | `O of [ `End ]
    | `P of [ `End ]
    | `Q of [ `End ]
    | `R of [ `End ]
    | `S of [ `End ]
    | `T of [ `End ]
    | `U of [ `End ]
    | `V of [ `End ]
    | `W of [ `End ]
    | `X of [ `End ]
    | `Y of [ `End ]
    | `Z of [ `End ] ]
= `A `End
- : [ `C of [ `End ] ] = `C `End
- : [ `C of [ `End ] ] = `C `End
- : [ `End ] = `End
- : [ `a of [ `End ] ] = `a `End
- : [ `abc of [ `End ] ] = `abc `End
- : [ `a of [ `End ] | `b of [ `End ] ] = `a `End
Characters 1-8:
  (`c `End :[%re"[ab]"]) ;; (*fail*)
   ^^^^^^^
Error: This expression has type [> `c of [> `End ] ]
       but an expression was expected of type
         [ `a of [ `End ] | `b of [ `End ] ]
       The second variant type does not allow tag(s) `c
- : [ `a of [ `End ] | `b of [ `End ] | `c of [ `End ] | `d of [ `End ] ] =
`a `End
Characters 1-8:
  (`x `End :[%re"[a-d]"]) ;; (*fail*)
   ^^^^^^^
Error: This expression has type [> `x of [> `End ] ]
       but an expression was expected of type
         [ `a of [ `End ] | `b of [ `End ] | `c of [ `End ] | `d of [ `End ] ]
       The second variant type does not allow tag(s) `x
- : [ `End | `a of 'a ] as 'a = `a (`a (`a `End))
- : [ `End | `a of 'a ] as 'a = `End
- : [ `a of [ `End | `a of 'a ] as 'a ] = `a (`a (`a `End))
Characters 1-5:
  (`End :[%re"a+"]) ;; (*fail*)
   ^^^^
Error: This expression has type [> `End ]
       but an expression was expected of type
         [ `a of [ `End | `a of 'a ] as 'a ]
       The second variant type does not allow tag(s) `End
- : [ `a of [ `a of [ `a of [ `End | `a of 'a ] as 'a ] ] ] =
`a (`a (`a `End))
- : [ `a of [ `a of [ `a of [ `End | `a of 'a ] as 'a ] ] ] =
`a (`a (`a (`a `End)))
Characters 8-12:
  (`a (`a `End) :[%re"a{3,}"]) ;; (*fail*)
          ^^^^
Error: This expression has type [> `End ]
       but an expression was expected of type
         [ `a of [ `End | `a of 'a ] as 'a ]
       The second variant type does not allow tag(s) `End
- : [ `a of [ `a of [ `a of [ `End | `a of [ `End ] ] ] ] ] =
`a (`a (`a `End))
- : [ `a of [ `a of [ `a of [ `End | `a of [ `End ] ] ] ] ] =
`a (`a (`a (`a `End)))
Characters 8-12:
  (`a (`a `End) :[%re"a{3,4}"]) ;; (*fail*)
          ^^^^
Error: This expression has type [> `End ]
       but an expression was expected of type
         [ `a of [ `End | `a of [ `End ] ] ]
       The second variant type does not allow tag(s) `End
Characters 16-25:
  (`a (`a (`a (`a (`a `End)))) :[%re"a{3,4}"]) ;; (*fail*)
                  ^^^^^^^^^
Error: This expression has type [> `a of [> `End ] ]
       but an expression was expected of type [ `End ]
       The second variant type does not allow tag(s) `a
- : [ `b of [ `End ] ] = `b `End
Characters 1-8:
  (`a `End :[%re"[ab]&[bc]"]) ;; (*fail*)
   ^^^^^^^
Error: This expression has type [> `a of [> `End ] ]
       but an expression was expected of type [ `b of [ `End ] ]
       The second variant type does not allow tag(s) `a
val id : 'a -> 'a = <fun>
- : [ `a of [ `a of [ `End | `a of 'a ] as 'a ] ] -> [ `a of [ `a of 'a ] ] =
<fun>
- : [ `b of [ `End ] ] -> [ `b of [ `End ] ] = <fun>
- : [ `a of [ `c of [ `End ] | `d of [ `End ] ]
    | `b of [ `c of [ `End ] | `d of [ `End ] ] ] ->
    [ `a of [ `c of [ `End ] | `d of [ `End ] ]
    | `b of [ `c of [ `End ] | `d of [ `End ] ] ]
= <fun>

val f : ([ `End | `a of 'a ] as 'a) -> ([ `End | `a of 'b | `b of 'b ] as 'b) =
   <fun>